### PR TITLE
[KubeRay][Autoscaler] Enable launching nodes in the main thread.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

This PR (RFC) is a further stability improvement for autoscaling with KubeRay. The new configuration introduced may be useful for other node providers as well. 

This PR adds a flag to enable creating nodes in the main thread in the autoscaler. The flag is turned on for the KubeRay node provider. KubeRay already uses a flag to disable node updater threads -- with the changes from this PR, the autoscaler becomes single-threaded when launching on KubeRay.

### Context
Currently, the autoscaler launches nodes in batches in a background thread. This makes sense when directly managing VMs on EC2 because the `node_provider.create_node` translates to `ec2_client.create_instance`, which is a blocking operation with non-negligible runtime and with potential to be throttled by the cloud provider.

For other node provider implementations (e.g. KubeRay) create_node is cheap (e.g. just translates to a write to etcd). So it doesn't make sense to batch in a background thread. If you launch in the background, it becomes harder to correctly implement NodeProviderInterface (e.g. you have to figure out thread safety). Managing the extra threads also negatively impacts the stability of the autoscaler. 

### Implementation detail
The class `NodeLauncher` inherits from `threading.Thread`. This PR factors out a new class `BaseNodeLauncher`, which carries most of the existing node launching logic and launches nodes in the main thread. `NodeLauncher` is then rewritten to inherit from `BaseNodeLauncher` and `threading.Thread`.

### Note
The existing functionality of `NodeLauncher` is not touched by this PR. This PR is a no-op for all node providers, except for KubeRay.

### Other
This PR also updates autoscaler image in the KubeRay example configs to include changes from https://github.com/ray-project/ray/pull/24590.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] New integration test logic
   - [x] Validated by existing KubeRay e2e test

### CODE REVIEW

File: python/ray/tests/test_autoscaler.py [hunk 1]
```diff
@@ -1765,13 +1829,32 @@ def testAggressiveAutoscaling(self):
...
+        # for being idle and instantly re-created due to resource demand!
+        lm.update(
+            worker_ip,
+            mock_raylet_id(),
+            {},
+            {},
+            {},
+            waiting_bundles=[],
+            infeasible_bundles=[],
+        )
```
Comment (DmitriGekhtman) [inline python/ray/tests/test_autoscaler.py:1857]: (The new flag exposed a race condition in the existing test logic.)

**Comment by DmitriGekhtman** [comment]
cc @clarng

**Review by wuisawesome — COMMENTED** [review]
I don't think I have full context of kuberay rn, but IIRC the original reason for the background thread was that the ec2 api has a long tail latency in the minutes so if this is really the best way of having safety in kuberay, then lgtm. 

Will leave it to @sriram-anyscale and @pcmoritz who have probably thought more about the kuberay implementation than me though.

**Comment by DmitriGekhtman** [comment]
> I don't think I have full context of kuberay rn, but IIRC the original reason for the background thread was that the ec2 api has a long tail latency in the minutes so if this is really the best way of having safety in kuberay, then lgtm.

Latency of create_node is not an issue for KubeRay. Some benefits of doing everything in a single thread:
- You can reason about the order in which the autoscaler issues remote calls
- Locks can be removed from the KuberayNodeProvider implementation.

In an ideal implementation of the node provider interface,
- create_node and terminate_node are cheap. When the method returns, you have a contract that a node will be created or deleted.
- there's read-after-write consistency: calling non_terminated_nodes immediately after create_node or terminate_node reflects the changes

**Comment by DmitriGekhtman** [comment]
Another way to interpret the suggested change:

The autoscaler was originally designed to take on full cluster management responsibility -- launching nodes efficiently is part of that, so it makes sense to have an apparatus to take care of that in that background.

For KubeRay, the bulk of cluster management is taken care of by the KubeRay operator and Kubernetes, and the autoscaler's only job is to communicate scale recommendations. `KuberayNodeProvider.create_node` is a cheap database write operation which can be handled in the main thread. The operator takes care of the asynchronous background launching.

File: python/ray/autoscaler/_private/autoscaler.py [hunk 2]
```diff
@@ -238,30 +242,55 @@ def read_fn():
         self.disable_node_updaters = self.config["provider"].get(
             "disable_node_updaters", False
         )
+        if self.disable_node_updaters:
```
Comment (sriram-anyscale) [inline python/ray/autoscaler/_private/autoscaler.py:245]: As a general comment it's a good practice to use the positive style for flags. People often get confused when they see "disableXXX == False" given that it is a double negative. Better to say "enableXXX". I realize you may not be able to make this change here - that is fine.

File: python/ray/autoscaler/_private/autoscaler.py [hunk 3]
```diff
@@ -238,30 +242,55 @@ def read_fn():
...
+            logger.info("Node updaters are disabled.")
```
Comment (sriram-anyscale) [inline python/ray/autoscaler/_private/autoscaler.py:246]: Why not log the flag value regardless of the value? Will it get too noisy?

**Review by sriram-anyscale — COMMENTED** [review]
It looks like the lock is still present in the Kuberay node_provider. Why not remove that?

**Comment by DmitriGekhtman** [comment]
> It looks like the lock is still present in the Kuberay node_provider. Why not remove that?

Good point -- I will do that in a follow-up. (Just keeping the diffs more incremental.)

File: python/ray/autoscaler/_private/autoscaler.py (hunk 3 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:246]: Will do.

**Comment by DmitriGekhtman** [comment]
> > It looks like the lock is still present in the Kuberay node_provider. Why not remove that?
> 
> Good point -- I will do that in a follow-up. (Just keeping the diffs more incremental.)

Actually, makes enough sense to do that in this PR. Will do that!

File: python/ray/autoscaler/_private/autoscaler.py [hunk 4]
```diff
@@ -238,30 +242,55 @@ def read_fn():
...
             "disable_launch_config_check", False
         )
+        if self.disable_launch_config_check:
+            logger.info("Launch config checks are disabled.")
+
+        # By default, the autoscaler launches nodes in batches asynchronously in a
+        # background thread.
+        # When the following flag is set, that behavior is disabled, so that nodes
+        # are launched in the main thread.
+        self.disable_background_launch_batch = self.config["provider"].get(
```
Comment (pcmoritz) [inline python/ray/autoscaler/_private/autoscaler.py:261]: Does this flag both deactivate the background thread launching and the batching? That would be good to clarify in the comment :)

File: python/ray/tests/test_autoscaler.py [hunk 5]
```diff
@@ -1494,7 +1524,11 @@ def testDynamicScaling5(self):
     def testDynamicScaling6(self):
         self.helperDynamicScaling(DrainNodeOutcome.FailedToFindIp)
 
-    def helperDynamicScaling(self, drain_node_outcome: DrainNodeOutcome):
+    def helperDynamicScaling(
+        self,
+        drain_node_outcome: DrainNodeOutcome = DrainNodeOutcome.Succeeded,
+        foreground_node_launcher: bool = False,
```
Comment (pcmoritz) [inline python/ray/tests/test_autoscaler.py:1530]: Does this flag need to get passed through in the 

```
self._helperDynamicScaling(mock_metrics, mock_node_info_stub)
```

call below? I feel like it probably needs to :)

File: python/ray/autoscaler/_private/autoscaler.py (hunk 4 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:261]: Yes, good point.

File: python/ray/tests/test_autoscaler.py (hunk 5 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:1530]: prrrobabbly, nice catch

File: python/ray/autoscaler/_private/autoscaler.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:248]: Renamed it "Foreground node launch."

File: python/ray/autoscaler/_private/autoscaler.py [hunk 6]
```diff
@@ -236,32 +243,69 @@ def read_fn():
...
         # meaningful node "type" to enforce.
         self.disable_launch_config_check = self.config["provider"].get(
-            "disable_launch_config_check", False
+            DISABLE_LAUNCH_CONFIG_CHECK_KEY, False
+        )
+        logger.info(
+            f"{DISABLE_LAUNCH_CONFIG_CHECK_KEY}:{self.disable_launch_config_check}"
         )
+        if self.disable_launch_config_check:
+            logger.info("Launch config checks are disabled.")
```
Comment (pcmoritz) [inline python/ray/autoscaler/_private/autoscaler.py:264]: This is already covered by the logging
```
logger.info(
            f"{DISABLE_LAUNCH_CONFIG_CHECK_KEY}:{self.disable_launch_config_check}"
        )
```
right? Same for the other logging statement above :)

**Review by pcmoritz — APPROVED** [review]

File: python/ray/autoscaler/_private/autoscaler.py (hunk 6 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/autoscaler.py:264]: Oh, good point. The English sentence doesn't really add much color for anyone.

File: python/ray/tests/test_autoscaler.py (hunk 5 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler.py:1530]: Fixed.

**Comment by DmitriGekhtman** [comment]
Tests look good -- will do a bit of manual testing as a final check.

**Pull request merged**